packages/caching/tests/ChangedFilesDetectorTest.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\Caching\Tests;
+
+use Rector\Caching\ChangedFilesDetector;
+use Rector\Core\Testing\PHPUnit\AbstractRectorTestCase;
+use Symplify\SmartFileSystem\SmartFileInfo;
+
+final class ChangedFilesDetectorTest extends AbstractRectorTestCase
+{
+    /**
+     * @var ChangedFilesDetector
+     */
+    private $changedFilesDetector;
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $this->changedFilesDetector = self::$container->get(ChangedFilesDetector::class);
+    }
+
+    protected function tearDown(): void
+    {
+        parent::tearDown();
+
+        $this->changedFilesDetector->clear();
+    }
+
+    public function testHasFileChanged(): void
+    {
+        $smartFileInfo = $this->getSmartFileInfo();
+
+        $this->assertTrue($this->changedFilesDetector->hasFileChanged($smartFileInfo));
+        $this->changedFilesDetector->addFileWithDependencies($smartFileInfo, []);
+        $this->assertFalse($this->changedFilesDetector->hasFileChanged($smartFileInfo));
+    }
+
+    public function testInvalidateFile(): void
+    {
+        $smartFileInfo = $this->getSmartFileInfo();
+
+        $this->changedFilesDetector->addFileWithDependencies($smartFileInfo, []);
+        $this->changedFilesDetector->invalidateFile($smartFileInfo);
+        $this->assertTrue($this->changedFilesDetector->hasFileChanged($smartFileInfo));
+    }
+
+    public function testGetDependentFileInfos(): void
+    {
+        $smartFileInfo = $this->getSmartFileInfo();
+        $dependantFile = $this->filePathToTest();
+
+        $this->changedFilesDetector->addFileWithDependencies($smartFileInfo, [$dependantFile]);
+        $dependantSmartFileInfos = $this->changedFilesDetector->getDependentFileInfos($smartFileInfo);
+
+        $this->assertCount(1, $dependantSmartFileInfos);
+        $this->assertSame($dependantFile, $dependantSmartFileInfos[0]->getPathname());
+    }
```

Comment:
There 3 methods are very difficult to read for me.
How can we extend it with more case in the future? 

What about using 1 method and 1 data provider?

Reply:
I don't understand, data provider is used when you want to run same test with multiple different sets of params, this is not the cases here. Can you make example how you want it to look like?

Reply:
>  data provider is used when you want to run same test with multiple different sets of params

Exactly!

This is what I have in mind:

```php
public function test(array $inputFiles, array $changedFiles, int $expectedInvalidateCount)
{
    // ...
}

public function provideData()
{
    
    yield [[`A.php`, `B.php`, `C.php`], [], 0];

    // A is alone
    yield [[`A.php`, `B.php`, `C.php`], ['A.php'], 1];

    // C extends B, so changed B, also invalidated C
    yield [[`A.php`, `B.php`, `C.php`], ['B.php'], 2];
}
```

Reply:
I still don't understand, this is just changed files detector it is not responsible for invalidation of dependent files.

Reply:
Changed.